Source/r_sky.c

```diff
@@ -67,6 +71,73 @@ void R_InitSkyMap (void)
   skytexturemid = 100*FRACUNIT;
 }
 
+static byte V_GetPaletteIndex(byte *palette, int r, int g, int b)
+{
+  byte best;
+  int best_diff, diff;
+  int i;
+
+  best = 0; best_diff = INT_MAX;
+
+  for (i = 0; i < 256; ++i)
+  {
+    diff = (r - palette[3 * i + 0]) * (r - palette[3 * i + 0])
+          + (g - palette[3 * i + 1]) * (g - palette[3 * i + 1])
+          + (b - palette[3 * i + 2]) * (b - palette[3 * i + 2]);
+
+    if (diff < best_diff)
+    {
+      best = i;
+      best_diff = diff;
+    }
+
+    if (diff == 0)
+    {
+      break;
+    }
+  }
+
+  return best;
+}
+
+byte R_SkyBlendColor(int tex)
```

Comment:
I am having my problems with this function, because it only caches the return value for the last requested sky texture. In theory you may have an unlimited number of MBF skies in a map, all visible at the same time. This means that this function will be called over and over again during each rendered frame, which will be a performance desaster. 

I am playing the thought of creating a dynamic texture->color lookup table and reallocate it on demand, so that each texture has its color determined only once. What do you think of this?

Reply:
I agree, in fact EE does something similar: https://github.com/team-eternity/eternity/blob/85d5105207edf856d664932c428099dffb228b69/source/r_sky.cpp#L198